lib/Conversion/ExportVerilog/ExportVerilog.cpp

```diff
@@ -2792,6 +2799,85 @@ void StmtEmitter::emitStatementExpression(Operation *op) {
   emitLocationInfoAndNewLine(emittedExprs);
 }
 
+LogicalResult StmtEmitter::emitDeclaration(Operation *op) {
+  // If spillWiresAtPrepare option is disabled, use the previous emission
+  // method.
+  // TODO: Once ExportVerilog simplification finished, remove this condition.
+  if (!state.options.spillWiresAtPrepare)
+    return emitNoop();
+
+  emitSVAttributes(op);
+  auto value = op->getResult(0);
+  SmallPtrSet<Operation *, 8> opsForLocation;
+  opsForLocation.insert(op);
+
+  // Emit the leading word, like 'wire', 'reg' or 'logic'.
+  auto type = value.getType();
+  auto word = getVerilogDeclWord(op, state.options);
+  if (!isZeroBitType(type)) {
+    indent() << word;
+    auto extraIndent = word.empty() ? 0 : 1;
+    os.indent(maxDeclNameWidth - word.size() + extraIndent);
+  } else {
+    indent() << "// Zero width: " << word << ' ';
+  }
+
+  SmallString<8> typeString;
+  // Convert the port's type to a string and measure it.
+  {
+    llvm::raw_svector_ostream stringStream(typeString);
+    emitter.printPackedType(stripUnpackedTypes(type), stringStream,
+                            op->getLoc());
+  }
+  // Emit the type.
+  os << typeString;
+  if (typeString.size() < maxTypeWidth)
+    os.indent(maxTypeWidth - typeString.size());
+
+  // Emit the name.
+  os << names.getName(value);
+
+  // Print out any array subscripts or other post-name stuff.
+  emitter.printUnpackedTypePostfix(type, os);
+
+  if (auto localparam = dyn_cast<LocalParamOp>(op)) {
+    os << " = ";
+    emitter.printParamValue(localparam.getValue(), os, [&]() {
+      return op->emitOpError("invalid localparam value");
+    });
+  }
+
+  // Try inlining wire assignments into declarations.
+  if (isa<WireOp>(op)) {
+    // If the wire has a single assignment located at next to the wire, we can
+    // inline the assignment.
+    AssignOp singleAssign = {};
+    if (llvm::all_of(op->getUsers(),
+                     [&](Operation *user) {
+                       if (hasSVAttributes(user))
+                         return false;
+
+                       if (auto assign = dyn_cast<AssignOp>(user)) {
+                         singleAssign = assign;
+                         // Check that the assignment is next to the wire.
+                         return assign == op->getNextNode();
```

Comment:
So I think this is conservative, which seems fine, and it's efficient. But wouldn't it be more general and also correct to check `assign.getBlock() == op->getBlock()`?

Reply:
Yes, `assign == op->getNextNode()` is a bit conservative but `assign.getBlock() == op->getBlock()` might produce incorrect results. For example, we cannot inline the assign for `%a` in the following IR:
```
%a = sv.wire
%b = sv.wire
sv.assign %a, (sv.read_inout %b)
```
PrepareForEmission internally transforms graph regions into SSACFG; we can inline an assign if a source value can be usable at the program point of wire declaration.

Reply:
I see, so the alternative would be something like `op->isBeforeInBlock(assign)`? Not sure the exact api offhand, but isBeforeInBlock exists somewhere. If it's not blowing up runtime, that might be one route to explore.

Reply:
> so the alternative would be something like op->isBeforeInBlock(assign)?

Thanks, yes `isBeforeInBlock` can be used here, but we have to check that every sub operation of `op` satisfies that conditions.

Reply:
yeah... i am fine to keep this as-is, and if we decide it is insufficient, we can generalize and have some sort of smart optimization to not make the runtime quadratic

Reply:
Yeah, I think it makes sense to keep the current version and improve if necessary in the future.